src/docfx/build/toc/parser/MarkdownTocMarkup.cs

```diff
@@ -0,0 +1,210 @@
+                if (item != null)
+                {
+                    items.Add(item);
+                    var currentLevel = headingBlocks[i].Level;
+                    if (i + 1 < headingBlocks.Length && headingBlocks[i + 1].Level > currentLevel)
+                    {
+                        if (headingBlocks[i + 1].Level - currentLevel > 1)
+                        {
+                            throw Errors.InvalidTocLevel(filePath, currentLevel, headingBlocks[i + 1].Level).ToException();
+                        }
+
+                        var (children, count) = ConvertTo(filePath, headingBlocks, i + 1);
+                        item.Items = children;
+                        i += count;
+                        childrenCount += count;
+                    }
+
+                    if (i + 1 < headingBlocks.Length && headingBlocks[i + 1].Level < currentLevel)
+                    {
+                        break;
+                    }
+                }
+            }
+            while (++i < headingBlocks.Length);
+
+            return (items, items.Count + childrenCount);
+
+            TableOfContentsInputItem GetItem(HeadingBlock block)
+            {
+                if (block.Inline == null || !block.Inline.Any())
+                {
+                    return null;
+                }
+
+                string name = null;
+                string displayName = null;
+                string href = null;
+
+                if (block.Inline.First() is LinkInline linkInline)
+                {
+                    if (!string.IsNullOrEmpty(linkInline.Url))
+                        href = linkInline.Url;
+                    if (!string.IsNullOrEmpty(linkInline.Title))
+                        displayName = linkInline.Title;
+
+                    if (linkInline.Any() && linkInline.First() is LiteralInline literalInline)
+                    {
+                        name = literalInline.Content.ToString();
+                    }
+                }
+                else if (block.Inline.First() is LiteralInline literalInline)
+                {
+                    name = literalInline.Content.ToString();
+                }
+
+                var currentItem = new TableOfContentsInputItem
+                {
+                    DisplayName = displayName,
+                    Name = name,
+                    Href = href,
+                };
+
+                return currentItem;
+            }
+        }
+
+        private static MarkdownPipeline CreateMarkdownPipeline()
+        {
+            var markdownContext = new MarkdownContext(null, LogWarning, LogError, null, null);
+
+            return new MarkdownPipelineBuilder()
+                .UseYamlFrontMatter()
+                .UseDocfxExtensions(markdownContext)
+                .UseExtractTocYamlHeader()
+                .Build();
+        }
+
+        private static void LogError(string code, string message, string doc, int line)
```

Comment:
>LogError [](start = 28, length = 8)

Can we put LogError and LogWarning in a central place? duplicated in Markup.cs

Reply:
because we are using some static filed in the method, it's hard to share

---
In reply to: [212504930](https://github.com/dotnet/docfx/pull/3269#discussion_r212504930) [](ancestors = 212504930)